File: ginivision/src/main/java/net/gini/android/vision/tracking/EventTracker.java
package net.gini.android.vision.tracking;

/**
 * Created by Alpar Szotyori on 03.03.2020.
 *
 * Copyright (c) 2020 Gini GmbH.
 */

import android.app.Activity;

import net.gini.android.vision.GiniVision;

/**
 * Implement this interface and pass it to the {@link GiniVision.Builder#setEventTracker(EventTracker)} to get informed about various events
 * occuring during the usage of the Gini Vision Library.
 *
 * <p> Check each event enum to find out which events are triggered on the different screens.
 *
 * <p> If you use the Screen API all events will be triggered automatically.
 *
 * <p> If you use the Component API some events will not be triggered (for ex. events which rely on {@link Activity#onBackPressed()}). You
 * need to check whether all the events you are interested in are triggered.
 */
public interface EventTracker {

    /**
     * Called when an event is triggered on the onboarding screen.
     *
     * <p> See {@link OnboardingScreenEvent} for possible events.
     *
     * @param event the onboarding screen event
     */
    void onOnboardingScreenEvent(Event<OnboardingScreenEvent> event);

    /**
     * Called when an event is triggered on the camera screen.
     *
     * <p> See {@link CameraScreenEvent} for possible events.
     *
     * @param event the camera screen event
     */
    void onCameraScreenEvent(Event<CameraScreenEvent> event);

    /**
     * Called when an event is triggered on the review screen.
     *
     * <p> See {@link ReviewScreenEvent} for possible events.
     *
     * @param event the review screen event
     */
    void onReviewScreenEvent(Event<ReviewScreenEvent> event);

    /**
     * Called when an event is triggered on the analysis screen.
     *
     * <p> See {@link AnalysisScreenEvent} for possible events.
     *
     * @param event the analysis screen event
     */
    void onAnalysisScreenEvent(Event<AnalysisScreenEvent> event);
}

File: ginivision/src/main/java/net/gini/android/vision/tracking/EventTracker.kt
package net.gini.android.vision.tracking

import net.gini.android.vision.GiniVision
import java.util.*

/**
 * Created by Alpar Szotyori on 27.02.2020.
 *
 * Copyright (c) 2020 Gini GmbH.
 */

interface EventTracker {
    fun onOnboardingScreenEvent(event: Event<OnboardingScreenEvent>)
    fun onCameraScreenEvent(event: Event<CameraScreenEvent>)
    fun onReviewScreenEvent(event: Event<ReviewScreenEvent>)
    fun onAnalysisScreenEvent(event: Event<AnalysisScreenEvent>)
}

class Event<T : Enum<T>> @JvmOverloads constructor(val type: T,
                                                   val details: Map<String, String> = emptyMap()) {

    override fun equals(other: Any?) = other is Event<*>
            && type == other.type
            && details == other.details

    override fun hashCode() = Objects.hash(type, details)

    override fun toString(): String {
        return "Event(type=$type, details=$details)"
    }
}

@JvmOverloads
internal fun trackOnboardingScreenEvent(event: OnboardingScreenEvent, details: Map<String, String> = emptyMap()) {
    if (GiniVision.hasInstance()) {
        GiniVision.getInstance().eventTracker.onOnboardingScreenEvent(Event(event, details))
    }
}

@JvmOverloads
internal fun trackCameraScreenEvent(event: CameraScreenEvent, details: Map<String, String> = emptyMap()) {
    if (GiniVision.hasInstance()) {
        GiniVision.getInstance().eventTracker.onCameraScreenEvent(Event(event, details))
    }
}

@JvmOverloads
internal fun trackReviewScreenEvent(event: ReviewScreenEvent, details: Map<String, String> = emptyMap()) {
    if (GiniVision.hasInstance()) {
        GiniVision.getInstance().eventTracker.onReviewScreenEvent(Event(event, details))
    }
}

@JvmOverloads
internal fun trackAnalysisScreenEvent(event: AnalysisScreenEvent, details: Map<String, String> = emptyMap()) {
    if (GiniVision.hasInstance()) {
        GiniVision.getInstance().eventTracker.onAnalysisScreenEvent(Event(event, details))
    }
}